### PR TITLE
picolibc: enable read()/write() with !VFS, implement missing fs functions

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

Picolibc does not implement `read()`/`write()` on it's own. We already have them implemented for the VFS adaption, but
the `read()`/`write()` functions still work without VFS when reading/writing to stdin, stdout or stderr.

This also provide the remaining functions necessary to build `tests/pkg_fatfs_vfs` with PICOLIBC.

### Testing procedure

```
make -C examples/gnrc_networking BOARD=samr21-xpro PICOLIBC=1 -j
```

works and the output of `ping6` looks normal.

```
make -C tests/pkg_fatfs_vfs BOARD=samr21-xpro PICOLIBC=1 -j
```

should now also work.

### Issues/PRs references

alternative to #14841

### CODE REVIEW

File: sys/picolibc_syscalls_default/syscalls.c [hunk 1]
```diff
@@ -282,13 +330,23 @@ int fcntl (int fd, int cmd, int arg)
...
+    int res = vfs_lseek(fd - VFS_FD_OFFSET, off, whence);
     if (res < 0) {
         /* vfs returns negative error codes */
         errno = -res;
         return -1;
     }
     return res;
+#else
+    (void) off;
+    (void) whence;
```
Comment (maribu) [inline sys/picolibc_syscalls_default/syscalls.c:348]:
```suggestion
    (void)off;
    (void)whence;
```

Just place those unconditionally at the start of the function; no need to guard them as they do no harm. More instances elsewhere.

File: sys/picolibc_syscalls_default/syscalls.c [hunk 2]
```diff
@@ -144,13 +150,20 @@ pid_t getpid(void)
...
 {
+#ifdef MODULE_VFS
     int fd = vfs_open(name, flags, mode);
     if (fd < 0) {
         /* vfs returns negative error codes */
         errno = -fd;
         return -1;
     }
-    return fd;
+#else
```
Comment (maribu) [inline sys/picolibc_syscalls_default/syscalls.c:160]:
```suggestion
    return fd + VFS_FS_OFFSET;
#else
```

File: sys/picolibc_syscalls_default/syscalls.c [hunk 3]
```diff
@@ -144,13 +150,20 @@ pid_t getpid(void)
...
+    (void) name;
+    (void) flags;
+    (void) mode;
```
Comment (maribu) [inline sys/picolibc_syscalls_default/syscalls.c:163]:
```suggestion
    (void)name;
    (void)flags;
    (void)mode;
```

**Review by maribu — COMMENTED** [review]
Some comments inline

File: sys/picolibc_syscalls_default/syscalls.c (hunk 1 above)
Comment (benpicco) [reply sys/picolibc_syscalls_default/syscalls.c:348]: Well they *do* catch errors (if one actually compiles with VFS :sweat_smile:)

File: sys/picolibc_syscalls_default/syscalls.c (hunk 2 above)
Comment (benpicco) [reply sys/picolibc_syscalls_default/syscalls.c:161]: Good catch!

**Comment by benpicco** [comment]
Uh guess it was late yesterday :sweat_smile: 
We do have [`vfs_bind_stdio()`](https://doc.riot-os.org/group__sys__vfs.html#ga9b760e38efd5f348b77e582954f0b141) which does exactly what it should.

So no need for any offsets then.
Still worth implementing those functions for the !VFS case though.

**Comment by fjmolinas** [comment]
It's not compiling for:

```
/data/riotbuild/riotbase/sys/picolibc_syscalls_default/syscalls.c:112:5: error: implicit declaration of function 'FDEV_SETUP_STREAM' [-Werror=implicit-function-declaration]
  112 |     FDEV_SETUP_STREAM(picolibc_put, picolibc_get, picolibc_flush, _FDEV_SETUP_RW);
      |     ^~~~~~~~~~~~~~~~~
```

**Comment by fjmolinas** [comment]
> It's not compiling for:

I'm probably missing something...

**Comment by benpicco** [comment]
What version of picolibc do you use / how did you install it?

**Comment by bergzand** [comment]
`tests/pkg_fatfs_vfs` compiles just fine here using the Docker image. The test appears to hang though (or maybe it takes just very long).

**Comment by fjmolinas** [comment]
> I'm probably missing something...

I allays forget to set the environment variable with `DOCKER_ENVIRONMENT_CMDLINE='-e PICOLIBC=1'`, works fine now, sorry for the noise

**Comment by benpicco** [comment]
I just installed the [Debian package](https://packages.debian.org/sid/picolibc-arm-none-eabi), no need for Docker :smile:

**Review by bergzand — APPROVED** [review]
Improves the current situation and only affects PicoLIBC. ACK!

**Review by fjmolinas — APPROVED** [review]
I ran `BOARD=nrf52840dk make -C tests/pkg_fatfs_vfs/ flash term -j` with and without `PICOLIBC`, I get the same result as on master, not sure if its the expected result, but its the same as in master.

```
2020-09-11 17:28:56,511 # main(): This is RIOT! (Version: 2020.10-devel-954-g3f3556-pr-14843)
2020-09-11 17:28:56,771 # Tests for FatFs over VFS - test results will be printed in the format test_name:result
2020-09-11 17:28:57,025 # test_mount__mount:[FAILED]
2020-09-11 17:28:57,027 # test_mount__umount:[FAILED]
2020-09-11 17:28:57,281 # test_open__mount:[FAILED]
2020-09-11 17:28:57,283 # test_open__open:[OK]
2020-09-11 17:28:57,285 # test_open__open_ro:[FAILED]
2020-09-11 17:28:57,288 # test_open__close_ro:[FAILED]
2020-09-11 17:28:57,290 # test_open__open_wo:[FAILED]
2020-09-11 17:28:57,293 # test_open__close_wo:[FAILED]
2020-09-11 17:28:57,295 # test_open__open_rw:[FAILED]
2020-09-11 17:28:57,298 # test_open__close_rw:[FAILED]
2020-09-11 17:28:57,300 # test_open__umount:[FAILED]
2020-09-11 17:28:57,555 # test_rw__mount:[FAILED]
2020-09-11 17:28:57,556 # test_rw__open_ro:[FAILED]
2020-09-11 17:28:57,559 # test_rw__read_ro:[FAILED]
2020-09-11 17:28:57,561 # test_rw__write_ro:[OK]
2020-09-11 17:28:57,563 # test_rw__close_ro:[FAILED]
2020-09-11 17:28:57,565 # test_rw__open_wo:[FAILED]
2020-09-11 17:28:57,567 # test_rw__read_wo:[OK]
2020-09-11 17:28:57,570 # test_rw__close_wo:[FAILED]
2020-09-11 17:28:57,572 # test_rw__open_rw:[FAILED]
2020-09-11 17:28:57,574 # test_rw__read_rw:[FAILED]
2020-09-11 17:28:57,577 # test_rw__write_rw:[FAILED]
2020-09-11 17:28:57,579 # test_rw__lseek:[FAILED]
2020-09-11 17:28:57,581 # test_rw__read_rw:[FAILED]
2020-09-11 17:28:57,583 # test_rw__close_rw:[FAILED]
2020-09-11 17:28:57,586 # test_rw__open_rwc:[FAILED]
2020-09-11 17:28:57,589 # test_rw__write_rwc:[FAILED]
2020-09-11 17:28:57,591 # test_rw__lseek_rwc:[FAILED]
2020-09-11 17:28:57,594 # test_rw__read_rwc:[FAILED]
2020-09-11 17:28:57,596 # test_rw__close_rwc:[FAILED]
2020-09-11 17:28:57,598 # test_rw__umount:[FAILED]
2020-09-11 17:28:57,852 # test_dir__mount:[FAILED]
2020-09-11 17:28:57,854 # test_dir__opendir:[FAILED]
2020-09-11 17:28:57,856 # test_dir__readdir1:[FAILED]
2020-09-11 17:28:57,859 # test_dir__readdir2:[FAILED]
2020-09-11 17:28:57,862 # test_dir__readdir_name:[FAILED]
2020-09-11 17:28:57,864 # test_dir__readdir3:[FAILED]
2020-09-11 17:28:57,867 # test_dir__closedir:[FAILED]
2020-09-11 17:28:57,869 # test_dir__umount:[FAILED]
2020-09-11 17:28:58,123 # test_rename__mount:[FAILED]
2020-09-11 17:28:58,126 # test_rename__rename:[FAILED]
2020-09-11 17:28:58,128 # test_rename__opendir:[FAILED]
2020-09-11 17:28:58,131 # test_rename__readdir1:[FAILED]
2020-09-11 17:28:58,134 # test_rename__readdir2:[FAILED]
2020-09-11 17:28:58,137 # test_rename__check_name:[FAILED]
2020-09-11 17:28:58,139 # test_rename__readdir3:[FAILED]
2020-09-11 17:28:58,142 # test_rename__closedir:[FAILED]
2020-09-11 17:28:58,145 # test_rename__umount:[FAILED]
2020-09-11 17:28:58,399 # test_unlink__mount:[FAILED]
2020-09-11 17:28:58,401 # test_unlink__unlink1:[FAILED]
2020-09-11 17:28:58,405 # test_unlink__unlink2:[FAILED]
2020-09-11 17:28:58,407 # test_unlink__opendir:[FAILED]
2020-09-11 17:28:58,409 # test_unlink__readdir:[FAILED]
2020-09-11 17:28:58,412 # test_unlink__closedir:[FAILED]
2020-09-11 17:28:58,414 # test_unlink__umount:[FAILED]
2020-09-11 17:28:58,670 # test_mkrmdir__mount:[FAILED]
2020-09-11 17:28:58,672 # test_mkrmdir__mkdir:[FAILED]
2020-09-11 17:28:58,675 # test_mkrmdir__opendir1:[FAILED]
2020-09-11 17:28:58,677 # test_mkrmdir__closedir:[FAILED]
2020-09-11 17:28:58,680 # test_mkrmdir__rmdir:[FAILED]
2020-09-11 17:28:58,682 # test_mkrmdir__opendir2:[OK]
2020-09-11 17:28:58,685 # test_mkrmdir__umount:[FAILED]
2020-09-11 17:28:58,939 # test_create__mount:[FAILED]
2020-09-11 17:28:58,941 # test_create__open_wo:[FAILED]
2020-09-11 17:28:58,944 # test_create__write_wo:[FAILED]
2020-09-11 17:28:58,947 # test_create__close_wo:[FAILED]
2020-09-11 17:28:58,950 # test_create__open_wo2:[FAILED]
2020-09-11 17:28:58,953 # test_create__write_wo2:[FAILED]
2020-09-11 17:28:58,955 # test_create__close_wo2:[FAILED]
2020-09-11 17:28:58,958 # test_create__umount:[FAILED]
2020-09-11 17:28:59,212 # test_stat__mount:[FAILED]
2020-09-11 17:28:59,214 # test_stat__open:[FAILED]
2020-09-11 17:28:59,216 # test_stat__write:[FAILED]
2020-09-11 17:28:59,219 # test_stat__close:[FAILED]
2020-09-11 17:28:59,221 # test_stat__open:[FAILED]
2020-09-11 17:28:59,223 # test_stat__stat:[FAILED]
2020-09-11 17:28:59,225 # test_stat__close:[FAILED]
2020-09-11 17:28:59,228 # test_stat__size:[FAILED]
2020-09-11 17:28:59,230 # test_stat__umount:[FAILED]
2020-09-11 17:28:59,484 # test_libc__mount:[FAILED]
2020-09-11 17:28:59,486 # test_libc__fopen:[OK]
2020-09-11 17:28:59,488 # test_libc__fopen_w:[FAILED]
2020-09-11 17:28:59,491 # test_libc__fopen_r:[FAILED]
2020-09-11 17:28:59,493 # test_libc__remove:[FAILED]
2020-09-11 17:28:59,495 # test_libc__fopen_a:[FAILED]
2020-09-11 17:28:59,498 # test_libc__fopen_a2:[FAILED]
2020-09-11 17:28:59,501 # test_libc__remove:[FAILED]
2020-09-11 17:28:59,503 # test_libc__umount:[FAILED]
2020-09-11 17:28:59,504 # Test end.
2020-09-11 17:29:07,393 # Exiting Pyterm

```

**Comment by benpicco** [comment]
Well the test uses `mtd_sdcard`, so unless you have an SD card wired up, it will fail.

**Pull request merged**